ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.engine.KsqlEngine;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.confluent.ksql.util.QueryMetadata;
+import java.net.URI;
+import java.net.URL;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.state.HostInfo;
+import org.apache.kafka.streams.state.StreamsMetadata;
+
+public final class HeartbeatHandler {
+
+  private final KsqlEngine engine;
+  private final ServiceContext serviceContext;
+  private final ConfigurationParameters configs;
+  private final ScheduledExecutorService scheduledExecutorService;
```

Comment:
It might be simpler to reason about but at the cost of serializing the three independent tasks. A server cannot send a heartbeat until it has processed all received heartbeats, etc. There is no reason in doing so , especially since there aren't many race conditions.
 
The race conditions that exist between the three different threads are:

1. `aliveHosts` is written to in `discoverCluster` and `processHeartbeat`. In `discoverCluster` it is set only the first time a server is discovered. It might happen that `processHeartbeat` sees it first, sets it to `false = down` and then `discoverCluster` sets it to true. This is not an issue as we made the decision to be optimistic when discovering servers for sake of availability. The other way around, is that `discoverCluster` will set it to true and then `processHeartbeat` will set it to false. Again, not an issue. This is the normal execution. For every next execution of `processHeartbeat` that modifies `aliveHosts`, it will be the only writer as the server has already been discovered.

2. `remoteHostsInfo` is written in `discoverCluster` and read in `processHeartbeat`. I use a `ConcurrentHashMap` so no problems here.

3. `remoteHostsUri` is written in `discoverCluster` and read in `sendHeartbeat`. Again, I use a `ConcurrentHashMap`.

For the latter two cases, there is no data corruption as there is only one writer at a time. The only thing that can happen is that an entry is added to the map, but the readers will not see it due to weak consistency. But this is not an error.

(I will look into whether I can simplify the design by using a single data structure for `remoteHosts` and `aliveHosts` which you mention in a later comment)

Reply:
I think what @agavra is talking about is more structuring this as an event loop with a https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/DelayQueue.html holding the events.. (ala Actor pattern). This would eliminate the need for concurrency/locking. 

Here is why I think it makes sense to keep them as runnables. Atleast one of the tasks does an external RPC, that could block for a long time say network is slow/socket is dead... In this scenario, the tasks will pile up behind it.. So, let's keep it as runnables.. We probably don't need 3 whole threads in the pool. We can crank the default down once we understand how it behaves in the real world

Reply:
On a related note, one thing to understand would be if sending heartbeat tasks, can starve the other two shorter tasks.. i.e can happen if a SendHeartbeat tasks hangs (due to reasons I mentioned above) and two more SendHeartbeat are scheduled..  IIRC It keeps one instance of a task inflight. but good to check

Reply:
I've never used a DelayQueue or heard of the Actor pattern by name, but that's essentially what I was going for. 

I think there are lots of issues that would be solved by using it, and I think making the RPCs non-blocking would solve the issues. Below are some races that I thought of. They may not be the worst, but it just shows the deep, deep reasoning that needs to happen and how fragile this code will be going forward if it's changed:

- if receivedHeartbeats is empty, we consider all hosts down. During startup, we will process heartbeats, but we haven't received any, so we set them all to be down, even though we wanted to eagerly consider them up?
- if receivedHeartbeats is non-empty, but doesn't cover some of the alive hosts, then we don't compute the heartbeats for those hosts (i.e. if we receive heartbeats from X, but not from Y at all, we don't even bother considering Y so we wont mark it down)
- we can get a new treeMap entry after calling processHeartbeats while we're blocked on a (different treeMap's) `synchronized` block. if the window moves during that period, we will unnecessarily mark a host as down because we don't see anything in that window for it (another false startup case)
- if the discover task is dead, we'll be sending heartbeats to a stale set of hosts (which might be okay, but it might also not be)
- the discover task never cleans up hosts, meaning that we might be sending heartbeats to hosts that are long gone

Looking at this list there's lots of issues, some that are simplified by the actor model, some that aren't and need to be fixed independently. If we make the `makeHeartbeatRequest` non-blocking (we don't wait for a response anyway) then I think we solve the problem that you described.

Reply:
>I've never used a DelayQueue or heard of the Actor pattern by name,

https://akka.io/ might be worth looking into, if you want to learn more

>making the RPCs non-blocking would solve the issues.

Of course, we want non-blocking RPCs even for pull queries themselves. IIUC all of our current RPCs are blocking. I don't understand jersey et al that well to think about how to do async I/O. if there is not a simple way, we can't pursue this anymore. @vpapavas do you have ideas?

>some that are simplified by the actor model, some that aren't and need to be fixed independently

I actually think a lot of the event ordering scenarios you list, are in both. Actor pattern simplifies thinking about it as a state machine, which is easier to program against.  Synchronization wise, I think it's fairly straight forward with the TreeMap level locking now. 


> I think there are lots of issues that would be solved by using it, and I think making the RPCs non-blocking would solve the issues.

you are proposing a fire-and-forget for heartbeats? Even that needs us to be handed a Future somehow using an async API right? 

>the deep, deep reasoning that needs to happen and how fragile this code will be going forward if it's changed:

I don't think locks are this scary. and Code is editable :) . So we can change it as well. 

@vpapavas Its worth thinking about all these cases once. But I will leave it you to make the call on switching to the actor model or not. or if you want to take more time to understand and reimplement in a followon pr.. your call.

Reply:
+1 on everything you say above, it all makes sense 

> you are proposing a fire-and-forget for heartbeats? Even that needs us to be handed a Future somehow using an async API right?

Do we need to handle the futures? If we did, we could wrap them in a callback if we really want to and pass an executor to them (ListenableFuture style) but today we don't handle the response from heartbeats anyway so why not just fire-and-forget?

Reply:
I was talking more about how we accomplish this in our current framework, given all I see is synchronous calls. Dug around a bit and I think we could make a async() request like below. It probably makes sense to make heartbeat message sending async anyway.. (regardless of what we decide on threading/looping).. It should be much make everything lot lighter..  

```
Future<Response> responseFuture = target
        .path(path)
        .request(MediaType.APPLICATION_JSON_TYPE)
        .property(ClientProperties.READ_TIMEOUT, readTimeoutMs.orElse(0))
        .headers(headers())
        .async()
        .post(Entity.json(jsonEntity));
```

Reply:
Let me see if I follow the discussion: There are three issues: 1) Lifecycle and error handling of the three independent runnables 2) Interplay of the three runnables between their shared data structures 3) Asynchronous RPC calls.

Regarding (1) @vinothchandar suggests using the DelayQueue : the benefits of it is that tasks will not get lost, if one errors out or crashes others can still be added to the queue and consumed? I don't understand what asynchronous RPCs have to do with the lifecycle management. Moreover, this point is orthogonal to the ones below, right? 

Regarding (2) , these are all design decisions. I will document the ones I made in the code and we can discuss about changing them but in the end, I feel we will have to see it in practice and decide what is the best course. 

>if receivedHeartbeats is empty, we consider all hosts down. During startup, we will process heartbeats, but we haven't received any, so we set them all to be down, even though we wanted to eagerly consider them up?

What is the alternative? Until when should a server be considered as eagerly up? Consider them alive until we have received at least one heartbeat from a server?  Order the runnables so that `processHeartbeat` runs after `sendHeartbeat` or with a delayed start time? 

>if receivedHeartbeats is non-empty, but doesn't cover some of the alive hosts, then we don't compute the heartbeats for those hosts (i.e. if we receive heartbeats from X, but not from Y at all, we don't even bother considering Y so we wont mark it down)

Correct, fixed.

>we can get a new treeMap entry after calling processHeartbeats while we're blocked on a (different treeMap's) synchronized block. if the window moves during that period, we will unnecessarily mark a host as down because we don't see anything in that window for it (another false startup case)

I don't follow here.

> if the discover task is dead, we'll be sending heartbeats to a stale set of hosts (which might be okay, but it might also not be)

If the task is dead, we will be sending heartbeats considering an old view of the cluster. Which I feel is better than having the entire heartbeat mechanism dying out.

>the discover task never cleans up hosts, meaning that we might be sending heartbeats to hosts that are long gone

There is no need to clean up hosts as we are sending heartbeats only to hosts that are alive.

Reply:
> I don't understand what asynchronous RPCs have to do with the lifecycle management. 
They are not directly related. But it came up here, because doing synchronous RPC was an impediment to the eventLoop based design..

Reply:
I changed the code to use `AbstractScheduledService` with error handling at startup/shutdown. I also implemented asynchronous RPC only for sending heartbeat.

>On a related note, one thing to understand would be if sending heartbeat tasks, can starve the other two shorter tasks.. i.e can happen if a SendHeartbeat tasks hangs (due to reasons I mentioned above) and two more SendHeartbeat are scheduled.. IIRC It keeps one instance of a task inflight. but good to check

From the javadocs: `runOneIteration`: Run one iteration of the scheduled task. If any invocation of this method throws an exception, the service will transition to the Service.State.FAILED state and this method will no longer be called.

I address this by adding try/catch for every service.

Moreover, from javadocs: If any execution of this task takes longer than its period, then subsequent executions may start late, but will not concurrently execute.

Given the above, what is missing from the lifecycle management? What are the benefits of using an event loop?

Reply:
lets file an issue for the event loop based design? (we can keep it in backlog for now) and revisit. What you did seems workable for now IMO

Reply:
Tracked by issue: #4233